pretty: try to align exprs at right of keyword, de-indent AND and OR

knz opened the pull request:
This PR adds a new option `Align` in `PrettyCfg` and makes good use of it to right-align SQL keywords if there is enough space on the line.

When enabled, the alignment is performed using a mix of tabs and spaces so that the output remains properly aligned even if the display tab width changes after the output has been generated (requested by @mjibson).

If the option is disabled, or there is not enough space on the line, the formatting uses "nested sections" as before.

For example, before or with `Align` disabled:

```
--------------------------------------------------
SELECT
    count(*) AS count,
    winner,
    counter * 60 * 5 AS counter
FROM
    (
        SELECT
            winner,
            round(length / 60 / 5) AS counter
        FROM players
        WHERE
            build = $1
            AND hero = $2
            OR region = $3
    )
GROUP BY winner, counter;
```

With `Align` enabled:

```
--------------------------------------------------
  SELECT count(*) AS count,
         winner,
         counter * 60 * 5 AS counter
    FROM (
            SELECT winner,
                   round(
                    length / 60 / 5
                   ) AS counter
              FROM players
             WHERE build = $1
               AND hero = $2
                OR region = $3
         )
GROUP BY winner, counter;
```

Note that tabs are still used whenever possible if the `UseTabs` option is set. The value of `UseTabs` has no bearing on the chosen layout.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/27260)
<!-- Reviewable:end -->

knz committed:
pretty: add support for 2-column tables with right-aligned 1st col

This patch adds support in the pretty library for groups of clauses
where the clause titles (e.g. SELECT, FROM, WHERE in SQL) are
right-aligned and the clause operands left-aligned under them.

For example:

```
SELECT aaa,
       bbb
  FROM ccc,
       ddd,
 WHERE eee
```

This alignment is tabwidth-agnostic (it always uses space characters
regardless of the `UseTabs` setting).

Release note: None

madelynnblue reviewed (commented):
Code LGTM, but I'm not sold on the formatting. Specifically:

1) The Align stuff assumes that a user's editor is displaying tabs at the same width as the pretty printer. When that doesn't happen, things like

```
SELECT 1,
    2
````

can occur where the `1` and `2` aren't visually aligned. If you follow this thought experiment to the end, it suggests we should just remove tabs altogether and use spaces. I think that *if* we decide we want this kind of alignment, that probably makes the most sense.

2) The right alignment removes a nice property we had, which was that all things at conceptual level `N` were indented the same. (For example, `SELECT a FROM t` would always put `a` and `t` at the same indentation.) This change removes that property, and I kinda liked it.

3) Regarding de-indenting AND and OR, this has the unfortunate side effect of bringing those operators to the same visual level (especially with right alignment) as WHERE, ORDER BY, FROM, etc. One of the purposes of indenting (with tabs or spaces) is to create a visual hierarchy. Deindenting some things weakens that since one can no longer immediately tell at which conceptual level a thing is only based on its initial indentation.

Overall: I like tabs but could be convinced to change. I like tabs because it's easy to deal with during manual editing. But if I'm not doing any more manual editing, spaces would be fine. For the others, I am open to suggestions about how to handle them, since they are quite aesthetic and could vary by person. Voting polls in our slack? Half serious.

knz reviewed (commented):
Thanks for your early feedback. I think there are two topics of discussion here, which I would like to group as follows:

A) one topic about the particular formatting choices (independent of the discussion about tabs vs spaces)

I made a mistake to show you this PR about alignment of expressions and negative nesting for AND/OR before I asked you to look at  #27264. As a matter of fact, in https://github.com/knz/prettier/ I did it the other way - first align the various SELECT sub-clauses, and only then align the operands and only then AND/OR.

The changes in #27264 resolve point (2) of your previous comment.

It does not really make sense to look at this PR before the other one, and I apologize for wasting your time.

Now, how to move forward with that discussion:

A.1) I will soon (perhaps that will wait until Friday) re-organize the PR to present the changes "in the right order" and weave #27264 into it. 
A.2) I will change the PR to make this alignment behavior configurable. (I believe we need to make this configurable so as to enable a discussion between parties to reveal where preferences stand.) Let's call this option `align` for future reference below.
A.3) we'll discuss the particular alignment of AND and OR, to resolve point (3) of your previous comment.

I have a specific argument, due to the way SQL logical planning works, why AND and OR should be negatively indented relative to their operands (I can explain later). Assuming the argument holds, I could however be convinced (maybe) that they can still be collectively indented relative to the sub-clause keyword. So perhaps we can solve your concern (3) with a middle ground around here:

```
WHERE     foo
      AND bar
      AND baz
```

or

```
WHERE
    foo
AND bar
AND baz
```


What do you think? But I suggest we defer that discussion to after points A.1 and A.2 have gotten somewhere.

Then 

B) discussion relative to tabs vs spaces

I'll want to start first with your remark:

> The Align stuff assumes that a user's editor is displaying tabs at the same width as the pretty printer.

To be honest at first I wasn't really comfortable with this remark. It seemed to me that **using tabs for pretty-printing /anything/ only makes sense if the tool used to view the result was using the same width as the pretty-printer**. This is inherently true of any system that /produces/ tabs for alignment. This is also the reason why most projects strongly favor using spaces for everything, since it's hard to guarantee that different people will use the same tab widths and the appearance of text/code files containing tabs is uncontrollable otherwise.

However, on second read, I tried harder to understand what you meant, to find out why you thought that it wasn't important before. Then I discovered that you had written your code to provide that property: it so appears that one can change the tab width *a posteriori* and still get perfectly aligned output given the current implementation of `doc()` methods. 

This property is, for example, important since most people will use a combination of text editors at 4 spaces per tab and terminals at 8 spaces per tab.

You obviously seem to care about that property a lot.

For the record, I do not care as much (I usually visualize my tabs at 8 spaces per tab even in my editors!), but I can acknowledge your concern here.

Meanwhile, I do fantastically care about the readability of SQL queries, and I think that the alignment discussed above in (A) is a very, very clear improvement overall for complex queries. 

So perhaps we can find a middle ground as follows:

- if the `align` parameter is set, then default `useTabs` to `false` always, so that the display is reliable everywhere by default.
- if `align` is not set, default `useTabs` to whatever we agree on later (I'd be ok with defaulting to `true`, but perhaps other people will have other opinions).

What do you think?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27260#-:-LGyn2a15bNcIe4Ptru8:b-278e4l)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mjibson commented:
My understanding of these various PRs is that #27264 should not be reviewed because it's going away and this PR will be fixed in the upcoming future, at which point I should review it again. Is that correct?

A.2) Regarding adding a new `align` option. This makes me sad because I think we're going in the wrong direction (I want fewer options, not more). However I'm ok with adding it if it encourages discussion about preferences and allows us to make opinionated decisions in the future.

A.3) I'm totally ok with

```
WHERE     foo
      AND bar
      AND baz
```

I tried to do something like that myself (but with `foo` on its own line) unsuccessfully.

B) Your "on second read" paragraph is accurate. I believe pretty strongly that if we do allow tabs, we should maintain the property that whatever the tab-width of the editor, it always produces visually aligned code.

If we are going to add an `align` option, I might merge it into the `useTabs` option as an enum with three values: `align`, `tabs`, `spaces`. Or maybe remove `spaces` altogether?

knz commented:
Yes your understanding is correct, I suggest to continue this discussion after next Friday when I rework both PRs.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

knz committed:
pretty: split nesting with tabs from nesting with spaces

This splits the Nest operator into NestT (nest with 1 tab) and
NestS (nest with some spaces). The rendering algorithm is then
modified to handle documents that are composed of "some amount of
tabs, then some amount of spaces, then the rest" instead of just "some
amount of spaces, then the rest".

The patch then modifies JoinGroup, Bracket, etc. to use NestT so that
they don't have to pass the indentation width any more.

This simplifies the user code and provides a very slight alloc boost:

```
name           old time/op    new time/op    delta
PrettyData-16     152ms ±30%     170ms ±29%    ~     (p=0.190 n=10+10)

name           old alloc/op   new alloc/op   delta
PrettyData-16    21.2MB ± 0%    21.2MB ± 0%  +0.09%  (p=0.000 n=9+10)

name           old allocs/op  new allocs/op  delta
PrettyData-16     6.34k ± 1%     6.27k ± 1%  -1.21%  (p=0.000 n=10+10)
```

Release note: None

knz committed:
pretty: create a tab-agnostic alignment operator

The new "Align" operator uses space-based padding to create
alignment. This should not be used directly and instead used
via `AlignUnder()` and `JoinGroupAligned()`.

- `AlignUnder` tries to use `Align` and, if that does not fit,
  "degrades" to `NestUnder`.

- `JoinGroupAligned` is analogous to `JoinGroup` except that it uses
  `AlignUnder` instead of `NestUnder`.

The use of `AlignUnder` would give us, for example:

input:
```
select k+v,k,(select 123)
```

output:

```
------------------ (available width)
SELECT k + v,
.......k,             -- filled to the left with spaces
.......(
>   >   SELECT 123    -- nested sub-item: cancel spaces, use tabs
.......)
```

```
----------- (available width)
SELECT
>   k + v,        -- filled to the left with a tab
>   k,
>   (
>   >   SELECT
>   >   >   123
>   )
```

This way, if the display device after pretty-printing uses a different
tab width than the one assumed by the algorithm, the display will
still make sense.

Release note: None

knz committed:
pretty: add an option to do alignmed formatting

This patch introduces a new option `Align` in `tree.PrettyCfg` to emit
in an "aligned" fashion. This mode is currently controlled by a
boolean option to enable discussion and comparison with/without.

When the option `Align` is enabled, the output tries to be aligned as
follows if there is enough space on the line:

```
SELECT a,
       b,
       c
```

This alignment is always realized with space characters, regardless of
the `UseTabs` option, so that variations in the physical tab with (on
the visualization device) do not break the alignment.

If there is not enough space on the line, or if `Align` is disabled,
the formatting automatically "degrades" to the previously defined
indented output:

```
SELECT
>   a,
>   b,
>   c
```

(using tab characters if `UseTabs` is set.)

Note that this is merely an intermediate patch because the resulting
output with `Align` is somewhat unsatisfactory: it is done separately
for every sub-clause in the SQL statement. As a result, the alignment
appears heterogeneous:

```
SELECT a,
       b,
       c
FROM d,
     e
```

A later commit will improve this situation.

Release note: None

knz committed:
pretty: add support for right-aligned SELECT clauses

This patch introduces support for right-aligning SELECT clauses.

When `Align` is set, before:

```
SELECT aaa,
       bbb
FROM ccc,
     ddd,
```

After:

```
SELECT aaa,
       bbb
  FROM ccc,
       ddd
```

To enable this change, additionally the patch changes the formatting
of the DISTINCT clause. Prior to this patch, the clause was always
printed to the right of the SELECT keyword. With this patch, the
DISTINCT clause, if any, is printed underneath with the target
expressions.

Release note: None

knz committed:
pretty: de-indent AND and OR

This patch adds a mode to de-indent AND and OR expressions, to follow
the convention used throughout the SQLite logic tests.

When `Align` is set, before:

```
WHERE aaa
      AND bbb
      AND ccc
```

After:

```
WHERE aaa
  AND bbb
  AND ccc
```

This de-indentation is not performed if there are not enough real
space characters on the left. This way, the result is stable even if
the display device has a different physical tab width than assumed by
the algorithm.

Release note: None

knz commented:
Note that this idea to de-indent AND and OR has a rather big precedent; it is used throughout the SQLite  test suite.

For example here: https://www.sqlite.org/sqllogictest/artifact/f9499fdbafa34e1e

knz committed:
pretty: make the Align option 4-valued

- value 0: don't align
- value 1: align clauses, don't de-indent AND/OR
- value 2: align clauses and de-indent AND/OR, like
  SQLite logic tests.
- value 3: special "matt" mode with aligned output
  but indented AND/OR clauses.

Release note: None

knz commented:
All right I have added the align option to do the AND/OR differently. You can test them with sqlfmt with the PR in https://github.com/mjibson/sqlfmt/pull/19.

knz committed:
pretty: ensure the document constants have type Doc

This simplifies usage in other parts of the code.

Release note: None

knz committed:
pretty: enable rows with no left label in rltables

This patch makes it possible to use a row with an empty cell on the
left side (= empty "label") in right-left aligned tables.

Release note: None

knz committed:
pretty: enable rows with nil right value in rltables

This patch ensures that the formatting of a row is correct even if the
right column is empty.

Release note: None

knz committed:
pretty: extend the interface of pretty.RLTable

Prior to this patch, there was redundant logic between
`(*PrettyCfg).rlTable()` and `pretty.RLTable()`. This patch adds a
boolean flag to the altter so that the former can be simplified.

Release note: None

knz committed:
pretty: format mutation statements and VALUES clauses also as tables

This extends the alignment strategu of SELECT to UPDATE, UPSERT,
INSERT, DELETE and VALUES.

Release note: None

knz committed:
pretty: simplify simplifyNil

The previous changes to format all the statements as tables has
ensured that Stack() never receives a `nil` reference as argument.

So `simplifyNil` does not need to test for this case any more. This
patch removes this check.

Release note: None

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27260)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/util/pretty/pretty_test.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27260#-LGtVuxoYjQFDjK4x_bu-r1-30:-LHSx2kc0NR8Sn0MVWlv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c26a4d8bf5e111702fd31314bd2c25444844ab53/pkg/util/pretty/pretty_test.go#L30)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I think `useTabs = false` would be a clearer example here because the example at 15 would be lined up regardless of editor configuration.
</blockquote></details>

Done.
___
*[pkg/util/pretty/util.go, line 222 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/27260#-LGtVuxoYjQFDjK4x_bv-r5-222:-LHSy97c84PuJKt139Ul:bovzcwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e5d298f5fb0950fd6c29d85fa811d6fb7ec951/pkg/util/pretty/util.go#L222)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I might use `append` instead of `items[i]` here and init items with `make([]Doc, 0, len(rows))` so that nil entries don't appear in the list. It happens to work because Stack calls ConcatLine which does the correct thing on nil, but it feels safer to be correct here.
</blockquote></details>

Done.

I do want to point out here that you have relied on this behavior of Stack() in other places: the pretty-printing of mutation statements. These were using `Stack(node.With.doc(p), ...)` assuming that things would work if the doc() method returned `nil`. 

I have added a commit to this PR to remove this requirement on `Stack()` (by updating the pretty-printing of mutations), then another to remove the special handling of `nil`.
___
*[pkg/util/pretty/util.go, line 223 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/27260#-LGtVuxoYjQFDjK4x_bv-r5-223:-LHSzMvRAejMNOMTEEvn:birs5p5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e5d298f5fb0950fd6c29d85fa811d6fb7ec951/pkg/util/pretty/util.go#L223)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

The commits here are not in correct order. The Align, JoinGroupAligned, and NestT functions don't exist yet. This came up because I wanted to check if Align already Group'd its return, making the call to Group here needless. I'm not sure how annoying it is to reorder them, but it might be helpful if it's doable.
</blockquote></details>

Are you sure? I have double-checked and every commit in turn passes all tests.

As to your question: no, `Align` does not group its return value (it shouldn't - that's not its purpose).


<!-- Sent from Reviewable.io -->

mjibson commented on pkg/sql/sem/tree/pretty_test.go:
```diff
@@ -47,13 +47,18 @@ func TestPrettyData(t *testing.T) {
 		runTestPrettyData(t, "align", cfg, matches)
 	})
+	cfg.Align = tree.PrettyAlignOnly
+	t.Run("align", func(t *testing.T) {
```
align -> `align2` probably?

mjibson commented on pkg/sql/sem/tree/pretty_test.go:
```diff
@@ -47,13 +47,18 @@ func TestPrettyData(t *testing.T) {
 	})
+	cfg.Align = tree.PrettyAlignOnly
+	t.Run("align", func(t *testing.T) {
+		runTestPrettyData(t, "align2", cfg, matches)
```
`align` and `align2` in the file names are a bit vague. Could we do something descriptive instead? Maybe `deindent` and `align`? Or `align-deindent` and `align-only`? Also I don't care too much so leave as is if you like.

knz replied:
Went for more descriptive names. Thanks for the suggestion.

knz committed:
pretty: use more descriptive test names

Release note: None

knz commented:
thank you!

bors r+
